File: automedal/agent/kernel.py
"""AgentKernel — the bespoke async tool-call loop.

This is the entire agent runtime: ~150 lines of pure orchestration.
No middleware, no graph, no framework. Just:

    while step < max_steps:
        turn = await provider.chat_stream(...)
        if not turn.tool_calls:
            return    # assistant produced final text
        for call in turn.tool_calls:
            result = await tool(**call.args)
            messages.append(tool_result for call.id)

The kernel is provider-agnostic: it speaks the internal Anthropic-flavored
message shape defined in `providers/base.py` and lets each provider adapter
translate at the wire. Tools are plain `Tool` dataclasses (see `tools/base.py`).
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass, field
from typing import Any

from automedal.agent.errors import format_error
from automedal.agent.providers.base import ChatProvider, ChatTurn, ToolCall, Usage
from automedal.agent.tools import Tool, ToolResult


@dataclass
class RunReport:
    """Summary of one kernel run."""
    stop: str                           # "assistant_done" | "max_steps" | "provider_error"
    final_text: str                     # last assistant text (or empty)
    messages: list[dict]                # full transcript including tool turns
    steps: int
    usage_total: Usage
    error: str | None = None


@dataclass
class AgentKernel:
    """One kernel = one phase invocation. Build a fresh kernel per phase."""

    provider: ChatProvider
    system_prompt: str
    tools: list[Tool]
    events: Any                                       # EventSink or None
    max_steps: int = 50
    parallel_tool_calls: bool = True                  # honor the model's batched tool_use blocks

    _tools_by_name: dict[str, Tool] = field(default_factory=dict, init=False)

    def __post_init__(self) -> None:
        self._tools_by_name = {t.name: t for t in self.tools}

    async def run(self, user_message: str) -> RunReport:
        messages: list[dict] = [{"role": "user", "content": user_message}]
        usage_total = Usage()

        for step in range(1, self.max_steps + 1):
            if self.events is not None:
                self.events.step_advance()

            try:
                turn: ChatTurn = await self.provider.chat_stream(
                    system=self.system_prompt,
                    messages=messages,
                    tools=self.tools,
                    events=self.events,
                )
            except Exception as exc:
                if self.events is not None:
                    self.events.error(where=f"provider.chat_stream step={step}", exc=exc)
                return RunReport(
                    stop="provider_error",
                    final_text="",
                    messages=messages,
                    steps=step,
                    usage_total=usage_total,
                    error=format_error(exc),
                )

            usage_total.in_tokens += turn.usage.in_tokens
            usage_total.out_tokens += turn.usage.out_tokens
            if self.events is not None and (turn.usage.in_tokens or turn.usage.out_tokens):
                self.events.usage(in_tokens=turn.usage.in_tokens, out_tokens=turn.usage.out_tokens)

            # Echo assistant blocks back into the transcript verbatim
            messages.append({"role": "assistant", "content": turn.assistant_blocks})

            if not turn.tool_calls:
                return RunReport(
                    stop="assistant_done",
                    final_text=turn.assistant_text,
                    messages=messages,
                    steps=step,
                    usage_total=usage_total,
                )

            # Execute tool calls — in parallel when the model batched several
            results = await self._execute_tools(turn.tool_calls)

            # Append a tool_result for every call, in the same order
            for call, result in zip(turn.tool_calls, results):
                messages.append({
                    "role": "tool",
                    "tool_use_id": call.id,
                    "content": result.text,
                    **({"is_error": True} if not result.ok else {}),
                })

        return RunReport(
            stop="max_steps",
            final_text="",
            messages=messages,
            steps=self.max_steps,
            usage_total=usage_total,
        )

    # ── tool execution ───────────────────────────────────────────────────────

    async def _execute_tools(self, calls: list[ToolCall]) -> list[ToolResult]:
        async def _one(call: ToolCall) -> ToolResult:
            tool = self._tools_by_name.get(call.name)
            if self.events is not None:
                self.events.tool_start(call_id=call.id, name=call.name, args=call.args)
            if tool is None:
                res = ToolResult(text=f"error: unknown tool {call.name!r}", ok=False)
            else:
                res = await tool(**call.args)
            if self.events is not None:
                self.events.tool_end(call_id=call.id, name=call.name, ok=res.ok, result=res.text)
            return res

        if self.parallel_tool_calls and len(calls) > 1:
            return await asyncio.gather(*(_one(c) for c in calls))
        return [await _one(c) for c in calls]

File: internal/agent/kernel.go
// AgentKernel — the bespoke tool-call loop.
//
// One kernel = one phase invocation. Build a fresh kernel per phase.
// The loop is provider-agnostic: it speaks the internal Anthropic-flavoured
// message shape (agent.Message + agent.Block) and lets each provider
// adapter translate at the wire.
//
// Port of automedal/agent/kernel.py:AgentKernel.
package agent

import (
	"context"
	"fmt"
	"sort"
	"strings"
	"sync"

	"github.com/Flameingmoy/automedal/internal/agent/tools"
)

// stop reasons the kernel surfaces
const (
	StopAssistantDone = "assistant_done"
	StopMaxSteps      = "max_steps"
	StopProviderError = "provider_error"
)

var lengthStopReasons = map[string]bool{
	"length":            true,
	"max_tokens":        true,
	"max_output_tokens": true,
}

// isLengthStop tracks the divergent stop_reason wording across providers.
func isLengthStop(s string) bool { return lengthStopReasons[strings.ToLower(s)] }

// RunReport summarises one kernel run.
type RunReport struct {
	Stop       string    // StopAssistantDone | StopMaxSteps | StopProviderError
	FinalText  string    // last assistant text (or empty)
	Messages   []Message // full transcript including tool turns
	Steps      int
	UsageTotal Usage
	Error      string // populated when Stop == StopProviderError
}

// ToolCall is one tool invocation requested by the model. Mirrors
// providers.ToolCall — duplicated here to avoid an import cycle.
type ToolCall struct {
	ID   string
	Name string
	Args map[string]any
}

// ChatTurn is what providers return. Duplicated from providers for the
// same reason as ToolCall — kernel cannot import providers.
type ChatTurn struct {
	AssistantBlocks []Block
	AssistantText   string
	ToolCalls       []ToolCall
	Usage           Usage
	StopReason      string
}

// ChatStreamFunc is the kernel's view of a provider — a single function
// closing over the underlying provider + its model. The cmd layer wires
// this from providers.ChatProvider.
type ChatStreamFunc func(
	ctx context.Context,
	system string,
	messages []Message,
	toolList []tools.Tool,
	events *EventSink,
) (*ChatTurn, error)

// AgentKernel is a single-phase tool-call loop.
type AgentKernel struct {
	Chat               ChatStreamFunc
	SystemPrompt       string
	Tools              []tools.Tool
	Events             *EventSink
	MaxSteps           int
	ParallelToolCalls  bool

	byName map[string]tools.Tool
}

// NewKernel constructs a kernel with sensible defaults.
func NewKernel(chat ChatStreamFunc, system string, ts []tools.Tool, events *EventSink) *AgentKernel {
	k := &AgentKernel{
		Chat:              chat,
		SystemPrompt:      system,
		Tools:             ts,
		Events:            events,
		MaxSteps:          50,
		ParallelToolCalls: true,
		byName:            make(map[string]tools.Tool, len(ts)),
	}
	for _, t := range ts {
		k.byName[t.Name] = t
	}
	return k
}

// Run executes the kernel against `userMessage`. Returns a RunReport.
func (k *AgentKernel) Run(ctx context.Context, userMessage string) RunReport {
	messages := []Message{{Role: "user", Content: userMessage}}
	usageTotal := Usage{}

	for step := 1; step <= k.MaxSteps; step++ {
		if k.Events != nil {
			k.Events.StepAdvance()
		}

		// Self-heal: patch any unanswered tool_use blocks left by an
		// interrupted prior step so the provider doesn't 400.
		if stubs := PatchDanglingToolCalls(&messages); stubs > 0 && k.Events != nil {
			k.Events.Notice("self_heal", fmt.Sprintf("patched %d dangling tool_use block(s)", stubs))
		}

		// Doom-loop guard: detect repetition / cycles in recent tool calls.
		if doom := CheckDoomLoop(messages); doom != "" {
			messages = append(messages, Message{Role: "user", Content: doom})
			if k.Events != nil {
				k.Events.Notice("doom_loop", doom)
			}
		}

		turn, err := k.Chat(ctx, k.SystemPrompt, messages, k.Tools, k.Events)
		if err != nil {
			if k.Events != nil {
				k.Events.Error(fmt.Sprintf("provider.chat_stream step=%d", step), err)
			}
			return RunReport{
				Stop:       StopProviderError,
				Messages:   messages,
				Steps:      step,
				UsageTotal: usageTotal,
				Error:      FormatError(err),
			}
		}

		usageTotal.InTokens += turn.Usage.InTokens
		usageTotal.OutTokens += turn.Usage.OutTokens
		if k.Events != nil && (turn.Usage.InTokens > 0 || turn.Usage.OutTokens > 0) {
			k.Events.Usage(turn.Usage.InTokens, turn.Usage.OutTokens)
		}

		// Truncation handler: model ran out of output budget mid-tool-call.
		// JSON arguments are garbage; drop the calls, keep any text prefix,
		// inject a hint, and re-loop.
		if isLengthStop(turn.StopReason) && len(turn.ToolCalls) > 0 {
			dropped := make([]string, 0, len(turn.ToolCalls))
			for _, tc := range turn.ToolCalls {
				dropped = append(dropped, tc.Name)
			}
			textBlocks := make([]Block, 0, len(turn.AssistantBlocks))
			for _, b := range turn.AssistantBlocks {
				if b.Type == "text" {
					textBlocks = append(textBlocks, b)
				}
			}
			messages = append(messages, Message{Role: "assistant", Content: textBlocks})
			hint := fmt.Sprintf(
				"Your previous response was truncated by the output token limit, so "+
					"the following tool calls were dropped: %v. Do NOT retry with the "+
					"same large content. For 'write_file' use bash with cat<<'HEREDOC', "+
					"or split into multiple smaller edit_file calls.",
				dropped,
			)
			messages = append(messages, Message{Role: "user", Content: hint})
			if k.Events != nil {
				k.Events.Notice("truncation", fmt.Sprintf("stop=length; dropped %v", dropped))
			}
			continue
		}

		// Echo assistant blocks back into the transcript verbatim.
		messages = append(messages, Message{Role: "assistant", Content: turn.AssistantBlocks})

		if len(turn.ToolCalls) == 0 {
			return RunReport{
				Stop:       StopAssistantDone,
				FinalText:  turn.AssistantText,
				Messages:   messages,
				Steps:      step,
				UsageTotal: usageTotal,
			}
		}

		results := k.executeTools(ctx, turn.ToolCalls)
		for i, call := range turn.ToolCalls {
			r := results[i]
			msg := Message{
				Role:      "tool",
				ToolUseID: call.ID,
				Content:   r.Text,
			}
			if !r.OK {
				msg.IsError = true
			}
			messages = append(messages, msg)
		}
	}

	return RunReport{
		Stop:       StopMaxSteps,
		Messages:   messages,
		Steps:      k.MaxSteps,
		UsageTotal: usageTotal,
	}
}

// executeTools dispatches every tool call (in parallel when batched) and
// returns results in the same order as `calls`.
func (k *AgentKernel) executeTools(ctx context.Context, calls []ToolCall) []tools.ToolResult {
	out := make([]tools.ToolResult, len(calls))

	dispatch := func(i int) {
		call := calls[i]
		if k.Events != nil {
			k.Events.ToolStart(call.ID, call.Name, call.Args)
		}
		t, ok := k.byName[call.Name]
		var res tools.ToolResult
		if !ok {
			res = tools.Error("error: unknown tool %q", call.Name)
		} else {
			res = t.Invoke(ctx, call.Args)
		}
		if k.Events != nil {
			k.Events.ToolEnd(call.ID, call.Name, res.OK, res.Text)
		}
		out[i] = res
	}

	if k.ParallelToolCalls && len(calls) > 1 {
		var wg sync.WaitGroup
		wg.Add(len(calls))
		// Stable order in `out`; goroutines write disjoint indices.
		idxs := make([]int, len(calls))
		for i := range calls {
			idxs[i] = i
		}
		sort.Ints(idxs)
		for _, i := range idxs {
			i := i
			go func() {
				defer wg.Done()
				dispatch(i)
			}()
		}
		wg.Wait()
		return out
	}
	for i := range calls {
		dispatch(i)
	}
	return out
}
